pkg/river/encoding/block.go

```diff
@@ -0,0 +1,86 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/grafana/agent/pkg/river/internal/rivertags"
+	"github.com/grafana/agent/pkg/river/internal/value"
+)
+
+// BlockField represents the json representation of a river block.
+type BlockField struct {
+	Field           `json:",omitempty"`
+	ID              string        `json:"id,omitempty"`
+	Label           string        `json:"label,omitempty"`
+	Body            []interface{} `json:"body,omitempty"`
+	attributeFields []*AttributeField
+	blockFields     []*BlockField
+}
+
+func newBlock(val value.Value, f rivertags.Field) (*BlockField, error) {
+	bf := &BlockField{}
+	return bf, bf.convertBlock(val, f)
+}
+
+// MarshalJSON implements json marshaller.
+func (bf *BlockField) MarshalJSON() ([]byte, error) {
+	bf.Body = make([]interface{}, 0)
+	for _, x := range bf.attributeFields {
+		bf.Body = append(bf.Body, x)
+	}
+	for _, x := range bf.blockFields {
+		bf.Body = append(bf.Body, x)
+	}
+	type temp BlockField
+	return json.Marshal((*temp)(bf))
+}
+
+func (bf *BlockField) isValid() bool {
+	if bf == nil {
+		return false
+	}
+	return len(bf.blockFields)+len(bf.attributeFields) > 0
+}
+
+func (bf *BlockField) convertBlock(val value.Value, f rivertags.Field) error {
+	if val.Reflect().Kind() != reflect.Struct {
+		return fmt.Errorf("convertBlock cannot work on interface or slices")
+	}
+	if val.Interface() == nil {
+		return nil
+	}
```

Comment:
Will this ever be true? Do you mean to check for `val.Type() == value.TypeNull` here?

Reply:
I wanted to handle the case: the two should be the equivalent but if value.TypeNull is easier to understand I will change to that.

Reply:
My reasoning here is that `val.Interface()` and `val.Reflect()` effectively convert the River value back to a Go value, which should be used extremely sparingly and only when it makes sense. 

If we're dealing with River values, we should be using the River API as much as we can.